ledger/eval/eval.go

```diff
@@ -788,13 +788,89 @@ func StartEvaluator(l LedgerForEvaluator, hdr bookkeeping.BlockHeader, evalOpts
 		return nil, fmt.Errorf("overflowed subtracting rewards for block %v", hdr.Round)
 	}
 
+	if eval.eligibleForIncentives(prevHeader.Proposer) {
+		incentive, _ := basics.NewPercent(proto.MiningPercent).DivvyAlgos(prevHeader.FeesCollected)
+		total, o := basics.OAddA(incentive, prevHeader.Bonus)
+		if o {
+			return nil, fmt.Errorf("overflowed adding bonus incentive %d %d", incentive, prevHeader.Bonus)
+		}
+
+		sink, err := eval.state.lookup(prevHeader.FeeSink)
+		if err != nil {
+			return nil, err
+		}
+		available := sink.AvailableBalance(&proto)
+		total = basics.MinA(total, available)
+
+		err = eval.state.Move(prevHeader.FeeSink, prevHeader.Proposer, total, nil, nil)
+		if err != nil {
+			// Should be impossible, we used AvailableBalance
+			return nil, fmt.Errorf("unable to pay block incentive: %v", err)
+		}
+	}
+
+	// Note their proposal if MiningEnabled.
+	if proto.EnableMining {
+		prp, err := eval.state.Get(prevHeader.Proposer, false)
+		if err != nil {
+			return nil, err
+		}
+		prp.LastProposed = hdr.Round - 1
+		// An account could propose, even while suspended, because of the 320 round lookback.
+		if prp.Suspended() {
+			prp.Status = basics.Online
+		}
+		err = eval.state.Put(prevHeader.Proposer, prp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	if eval.Tracer != nil {
 		eval.Tracer.BeforeBlock(&eval.block.BlockHeader)
 	}
 
 	return eval, nil
 }
 
+var (
+	// these would become ConsensusParameters if we ever wanted to change them
+
+	// incentiveMinBalance is the minimum balance an account must have to be
+	// eligible for incentives. It serves a couple purposes. First, it sets a
+	// manageable upper bound on the total number of incentivized participating
+	// accounts.  This means it is possible to track these accounts in memory,
+	// which is required to ensure absenteeism checking time is bounded. Second,
+	// it ensures that smaller accounts continue to operate for the same
+	// motivations they had before block incentives were introduced. Without
+	// that assurance, it is difficult to model their behaviour - might many
+	// participants join for the hope of easy financial rewards, but without
+	// caring enough to run a high-quality node?
+	incentiveMinBalance = basics.Algos(100_000)
+
+	// incentiveMaxBalance is the maximum balance an account might have to be
+	// eligible for incentives. It encourages large accounts to split their
+	// stake to add resilience to consensus in the case of outages.  Of course,
+	// nothing in protocol can prevent such accounts from running nodes that
+	// share fate (same machine, same data center, etc), but this serves as a
+	// gentle reminder.
+	incentiveMaxBalance = basics.Algos(100_000_000)
+)
+
+func (eval *BlockEvaluator) eligibleForIncentives(proposer basics.Address) bool {
+	proposerState, err := eval.state.Get(proposer, true)
+	if err != nil {
+		return false
```

Comment:
What's your preference, to bail out on evaluating the block entirely? As a practical matter, an error here is probably so catastrophic that our decision probably doesn't matter. (Another error is about to happen.)

If we bail, we _can't_ propose or validate. If we proceed in the face of an error, the worst that happens (in the extremely unlikely event that the error is limited to that account) is that we think they're not eligible, but they are, so we disagree with the world. If we bail, we _definitely_ disagree with the world (by not proposing or validating). So I don't see any advantage to the two-return two-step, just to bail.

Reply:
I'm somewhat more worried now compared to when I wrote that.  If disagreeing with the world showed up in the block, it would be as I said. But I think it would only be in the StateDeltas, so it would be silent disagreement between nodes.

Reply:
This function is only called by `StartEvaluator`, which has an error return value. I'd feel more comfortable if this function bubbled up the error so that the procedure could fail in the normal way (at least with more likelihood of a good/clear log message).